### DISCUSSION TITLE
Typescript: props.value doesn't reflect isMulti config when passing component through HOC – bug?

### DISCUSSION THREAD

**Discussion by dakur** [JedWatson/react-select#4952] [Q&A]
Hi,

when I want to pass `Creatable` through HOC, it resolves `props.value` inside the HOC to `PropsValue` which is correct, but it doesn't take `isMulti` into account.

When I change this line in [react-select `types.ts`](https://github.com/JedWatson/react-select/blob/react-select%405.2.1/packages/react-select/src/types.ts#L18) and update it also in used places, it ~~doesn't throw errors anymore~~ throws different errors.

```ts
// before
export type PropsValue<Option> = MultiValue<Option> | SingleValue<Option>;
// after
export declare type PropsValue<Option, IsMulti extends boolean> = IsMulti extends true ? MultiValue<Option> : SingleValue<Option>;
```

Still, not sure if this is an error or my bad. Can you please look at it? Or should I rather open an issue? (didn't want to pollute the issue space when not clearly sure it is a bug)

I've [prepared a repo](https://github.com/dakur/react-select-playground/tree/props-value-in-hoc):
1. clone
1. `yarn start`
1. see TS errors

**Comment by Methuselah96** [comment]
Thanks for the detailed question!

The reason it doesn't take `isMulti` into account is because it's technically valid to pass an array with a single value into `value` when `isMulti` is `false`. I can't remember whether this was a conscious design choice or just an implementation detail that leaked into the types. I'd be happy to re-investigate that if you want me to.

Can you point me to how you're using the `value` prop in a way where this is causing a problem?

**Reply by dakur** [reply]
Thanks for you answer!

I understand the thing about different value type when `isMulti` is on/off. But as I [set `IsMulti` in type generic to `true`](https://github.com/dakur/react-select-playground/blob/props-value-in-hoc/src/CreatableSelectProps.ts#L9), I would expect it makes its way to `props.value` as well – which doesn't happen.

Here is a [repo](https://github.com/dakur/react-select-playground/tree/props-value-in-hoc) (already linked above) with real-world example: https://github.com/dakur/react-select-playground/blob/props-value-in-hoc/src/hoc/makeOptionIfCommaPressed.tsx#L23  
Point of this HOC is that if you type something into the creatable and press comma, the already input text should make itself an option. This is used for e-mail address input:

![image](https://user-images.githubusercontent.com/9862581/145588089-fb6c4583-e560-4433-a191-745dbe5e8717.png)

If I press comma, it transfers into option:

![image](https://user-images.githubusercontent.com/9862581/145588021-e8f6726b-5af1-4719-a97c-ff43ce62cb54.png)

In the repo, I replaced e-mail addresses with nicknames for technical simpleness, but the point is still the same.

Feel free to ask further question if it is not clear and thank you once more!
